src/resource.py

```diff
@@ -0,0 +1,142 @@
+        print('Done downloading.')
+
+        print(f'Saving file {filename} ...')
+        with open(destination_path, 'wb') as f:
+            f.write(file.content)
+        print('Done. Saved to: ' + destination_path)
+
+    @staticmethod
+    def _download_folder(file_url, destination_path, update_handling):
+        print('Downloading folder...')
+        folder_soup = BeautifulSoup(globals.global_session.get(file_url).content, 'html.parser')  # Get folder page
+        dir_name = folder_soup.find('div', role='main').find('h2').contents[0]  # Find folder title
+        folder_path = os.path.join(destination_path, dir_name)
+        if not os.path.exists(folder_path):
+            print(f'Creating directory: {dir_name} in {destination_path}')
+            os.mkdir(folder_path)
+        files = folder_soup.find_all('span', class_='fp-filename')  # Finds all files in folder page
+        for file in files:
+            if len(file.contents) < 1:
+                continue
+            file_url = file.parent['href']
+            Resource._download_file(file_url, folder_path, update_handling)
+
+    @staticmethod
+    def _download_assignment(file_url, destination_path, update_handling):
+        # TODO: possibly extract multiple files
+        print('Extracting file from assignment...')
+        assignment_soup = BeautifulSoup(globals.global_session.get(file_url).content, 'html.parser')   # Get assignment page
+        file_anchor = assignment_soup.find('div', class_='fileuploadsubmission').find('a')
+        if len(file_anchor.contents) < 1:
+            print('No file found')
+            return
+        file_url = file_anchor['href']
+        Resource._download_file(file_url, destination_path, update_handling)
+
+    def download(self, destination_dir, update_handling):
+        if not os.path.exists(destination_dir):
+            print(destination_dir + ' not found. Creating path: ' + destination_dir)
+            try:
+                # Create path (recursively)
+                os.makedirs(destination_dir)
+            except FileNotFoundError:
+                print(f'Could not create path {destination_dir}. Please check the path and try again.')
+                return
+        # TODO: check, check if resource is actually available for the user
+        #  (see: https://github.com/NewLordVile/tum-moodle-downloader/issues/11)
+        print(f"Attempting to download resource {self.name} with type {self.type} ...")
+        if self.type == 'file':
+            Resource._download_file(self.resource_url, destination_dir, update_handling)
+        elif self.type == 'folder':
+            Resource._download_folder(self.resource_url, destination_dir, update_handling)
+        elif self.type == 'assignment':
+            Resource._download_assignment(self.resource_url, destination_dir, update_handling)
+        else:
+            print(f"Cannot download resource '{self.name}': type '{self.type}' is not supported!")
```

Comment:
If you want, you can run a formatter to add the missing new lines to the files.

Reply:
Good idea! Any recommendations for formatters?

Reply:
PyCharm has one built right in. Ctrl+Alt+L, I believe.

Reply:
Thanks :)
One just has to be careful to tell it to use spaces rather than tabs